cpp/src/arrow/engine/substrait/type_internal.cc

```diff
@@ -273,25 +298,60 @@ struct DataTypeToProtoImpl {
   Status Visit(const Date32Type& t) {
     return SetWith(&substrait::Type::set_allocated_date);
   }
-  Status Visit(const Date64Type& t) { return NotImplemented(t); }
+  Status Visit(const Date64Type& t) { return EncodeUserDefined(t); }
 
   Status Visit(const TimestampType& t) {
-    if (t.unit() != TimeUnit::MICRO) return NotImplemented(t);
-
     if (t.timezone() == "") {
-      return SetWith(&substrait::Type::set_allocated_timestamp);
-    }
-    if (t.timezone() == TimestampTzTimezoneString()) {
-      return SetWith(&substrait::Type::set_allocated_timestamp_tz);
+      auto ts = SetWithThen(&substrait::Type::set_allocated_precision_timestamp);
```

Comment:
Looks like you could avoid duplicating the `switch` by:

```cpp
auto ts = SetWithThen(t.timezone() == "" ? &substrait::Type::set_allocated_precision_timestamp : &&substrait::Type::set_allocated_precision_timestamp_tz);
      switch (t.unit()) {
        case TimeUnit::SECOND:
          ts->set_precision(0);
          break;
```

Reply:
The two methods take a different argument (e.g. `foo(int x)` and `bar(double y)`) and so the two branches of the ternary operator don't resolve to the "same type".  However, I was able to collapse the switch statements using a templated helper method.

Reply:
A-ha. Hidden type conversions at play here :)